### PR TITLE
Broker command line options and build executable

### PR DESCRIPTION
**Pull request by kleunen**
I added command line options (and config file) to configure the broker. 

````
General options:
  --help                                produce help message
  --cfg arg                             Load configuration file
  --verbose arg (=1)                    set verbose level, possible values:
                                         0 - Fatal
                                         1 - Error
                                         2 - Warning
                                         3 - Info
                                         4 - Debug
                                         5 - Trace

TCP Server options:
  --tcp.port arg (=1883)                default port (TCP)
  --tcp.enable arg (=1)                 0/1

TCP Websocket Server options:
  --ws.port arg (=10080)                default port (TCP)
  --ws.enable arg (=0)                  0/1

TLS Server options:
  --tls.port arg (=8883)                default port (TLS)
  --tls.enable arg (=0)                 0/1
  --tls.certificate arg (=server.crt.pem)
                                        Certificate file
  --tls.private_key arg (=server.key.pem)
                                        Private key file

TLS Websocket Server options:
  --tlsws.port arg (=10443)             default port (TLS)
  --tlsws.enable arg (=0)               0/1
````

### CODE REVIEW

**Comment by redboltz** [comment]
I'm not sure why it happened but your PR seems to contain already merged change.
See
https://github.com/redboltz/mqtt_cpp/commit/aaa61d1378b4b423a638c6e713680a9ee9698fef

For example

https://github.com/redboltz/mqtt_cpp/pull/772/files#diff-8438e6574a9cbca06fc27190e2dab4239f23e5e53b736428ad7b1d37dec831e7R329

is 

https://github.com/redboltz/mqtt_cpp/commit/aaa61d1378b4b423a638c6e713680a9ee9698fef#diff-8438e6574a9cbca06fc27190e2dab4239f23e5e53b736428ad7b1d37dec831e7R329

Rebase might solve the problem....?

**Comment by redboltz** [comment]
Basically, I agree to add commandline options to the example broker.
Using boost program_options is good.
Providing full options only is good. I mean shortcut option like `-P` shouldn't be provided at least this timing. 

```
  --tcp.port arg (=1883)                default port (TCP)
  --tcp.enable arg (=1)                 0/1
```

But I don't think enabler and port pattern is good.
`tcp.port` depends on `tcp.enable`. It is confusing.
example/broker is just an example. So you don't need to care about backward compatibility.

I think that the better approach is only if `---tcp.port=1883` is provided then enable tcp.port, otherwise, disable tcp.port.
I know the well known MQTT (TCP) port is 1883. It is a good value for default.
So, how about this approach ?

Set the default `--cfg` parameter as `broker.conf`.
And mqtt_cpp provides the default `broker.conf`. In the `broker.conf`, set `tcp.port` as 1883.

By the way, if both `broker.conf` and commandline options are provided, then the commandline options should override the `broker.conf` one. Off course, the override should happen the individual option.

**Comment by kleunen** [comment]
> Rebase might solve the problem....?

Yes. Made small mistake there with rebase

**Comment by kleunen** [comment]
What do you think about the executable build by CI now? I think it is useful to add to downloads on release, but maybe only build on tagged commit?

**Comment by kleunen** [comment]
> But I don't think enabler and port pattern is good.

But then what if you want to enable tcp and use default port

--tcp.port
Uses default port 1883
--tcp.port=12345
Use specified port?

**Comment by kleunen** [comment]
And what about naming tls and tlsws or mqqts and wss?

**Comment by kleunen** [comment]
The branch is now properly based

**Comment by kleunen** [comment]
Static linking does give a warning, so I am not sure if it is really possible to completely static link the executable: 

````
/usr/bin/clang++ -DBOOST_ASIO_NO_DEPRECATED -DBOOST_MULTI_INDEX_DISABLE_SERIALIZATION -O3 -DNDEBUG -static -static-libgcc -static-libstdc++ CMakeFiles/broker.dir/broker.cpp.o -o broker  /opt/hostedtoolcache/boost/1.72.0/x64/lib/libboost_log-mt-x64.a /opt/hostedtoolcache/boost/1.72.0/x64/lib/libboost_program_options-mt-x64.a /opt/hostedtoolcache/boost/1.72.0/x64/lib/libboost_system-mt-x64.a /usr/lib/x86_64-linux-gnu/libz.a /usr/lib/x86_64-linux-gnu/libssl.a /usr/lib/x86_64-linux-gnu/libcrypto.a -ldl /opt/hostedtoolcache/boost/1.72.0/x64/lib/libboost_date_time-mt-x64.a /opt/hostedtoolcache/boost/1.72.0/x64/lib/libboost_chrono-mt-x64.a /opt/hostedtoolcache/boost/1.72.0/x64/lib/libboost_filesystem-mt-x64.a /opt/hostedtoolcache/boost/1.72.0/x64/lib/libboost_regex-mt-x64.a /opt/hostedtoolcache/boost/1.72.0/x64/lib/libboost_thread-mt-x64.a -lpthread /opt/hostedtoolcache/boost/1.72.0/x64/lib/libboost_atomic-mt-x64.a 
/usr/lib/x86_64-linux-gnu/libcrypto.a(dso_dlfcn.o): In function `dlfcn_globallookup':
(.text+0x11): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/x86_64-linux-gnu/libcrypto.a(b_addr.o): In function `BIO_lookup_ex':
(.text+0xcfc): warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/x86_64-linux-gnu/libcrypto.a(b_sock.o): In function `BIO_gethostbyname':
(.text+0x71): warning: Using 'gethostbyname' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking

**Comment by kleunen** [comment]
I removed the changes to CI from this PR, I will make a separate PR for this

File: CMakeLists.txt [hunk 1]
```diff
@@ -123,16 +123,25 @@ FIND_PACKAGE (Threads REQUIRED)
...
     MESSAGE (STATUS "Logging disabled")
-    FIND_PACKAGE (Boost 1.67.0 REQUIRED COMPONENTS system date_time)
+    FIND_PACKAGE (Boost 1.67.0 REQUIRED COMPONENTS system date_time program_options)
 ENDIF ()
 
 IF (MQTT_USE_TLS)
     FIND_PACKAGE (OpenSSL REQUIRED)
     IF (MQTT_USE_STATIC_OPENSSL)
-        FIND_PACKAGE (ZLIB REQUIRED)
+         IF (UNIX AND NOT APPLE)
```
Comment (kleunen) [inline CMakeLists.txt:135]: This is also related to making a static build with CI, so I should move this to separate PR

**Comment by ogis-kondo** [comment]
> > But I don't think enabler and port pattern is good.
> 
> But then what if you want to enable tcp and use default port
> 
> --tcp.port
> Uses default port 1883
> --tcp.port=12345
> Use specified port?
> 
> Or maybe:
> --tcp and --tcp.port as option ?

I mean if `--tcp.port` is not passed, then disable tcp port.
See https://wandbox.org/permlink/WVoBGEGkaHOWv4xd

If user writes `broker.conf` as follows:

```
[tcp]
port=5678
```

Use 5678 as TCP port,
See https://wandbox.org/permlink/rpFpFb7jsg36vAL8
Be careful, there is `broker.conf` tab on wandbox.

If user set command line option, then it is used:
See https://wandbox.org/permlink/NIC74audyQE2ImMl
Be careful, there is Runtime options: text box on the left of the code.


And we can provide the default `broker.conf` in example directory.
It can be copied to build directory using cmake COPY operation.

**Comment by redboltz** [comment]
> > But I don't think enabler and port pattern is good.
> 
> But then what if you want to enable tcp and use default port
> 
> --tcp.port
> Uses default port 1883
> --tcp.port=12345
> Use specified port?
> 
> Or maybe:
> --tcp and --tcp.port as option ?

I mean if `--tcp.port` is not passed, then disable tcp port.
See https://wandbox.org/permlink/WVoBGEGkaHOWv4xd

If user writes `broker.conf` as follows:

```
[tcp]
port=5678
```

Use 5678 as TCP port,
See https://wandbox.org/permlink/rpFpFb7jsg36vAL8
Be careful, there is `broker.conf` tab on wandbox.

If user set command line option, then it is used:
See https://wandbox.org/permlink/NIC74audyQE2ImMl
Be careful, there is Runtime options: text box on the left of the code.


And we can provide the default `broker.conf` in example directory.
It can be copied to build directory using cmake COPY operation.

**Comment by kleunen** [comment]
> If user set command line option, then it is used:
> See https://wandbox.org/permlink/NIC74audyQE2ImMl
> Be careful, there is Runtime options: text box on the left of the code.

But this means you always have to specify the port, to enable TCP protocol

**Comment by redboltz** [comment]
> > If user set command line option, then it is used:
> > See https://wandbox.org/permlink/NIC74audyQE2ImMl
> > Be careful, there is Runtime options: text box on the left of the code.
> 
> But this means you always have to specify the port, to enable TCP protocol

Yes. It is no problem.
User should provide TCP port every time. It is better than unexpected port opens silently.
If user don't want to provide command line options, user can use `broker.conf`.
And `broker.conf` is provided by default.
So user can simple run `./broker` then port 1883 is opened as TCP port.

**Comment by redboltz** [comment]
> And what about naming tls and tlsws or mqqts and wss?

You mean this ?

||TCP|WS|
|---|---|---|
|NON TLS|mqtt|ws|
|TLS|mqtts|wss|

I know this is widely used but I don't think it is good name for options.
Because it is not well structured.
ws and wss are used for generic websocket, not only mqtt.
I think that it is similar level as tcp. 

Here is the straight forward notation that reflects protocol layer:

||TCP|WS|
|---|---|---|
|NON TLS|tcp.port|tcp.ws.port|
|TLS|tcp.tls.port|tcp.tls.ws.port|

I think that it is too complicated.


How about this?
||TCP|WS|
|---|---|---|
|NON TLS|tcp.port|ws.port|
|TLS|tls.port|wss.port|

There is no `mqtt` information. It is redundant because the broker is for mqtt.
tcp, tls, ws, and wss are similar level.

**Comment by redboltz** [comment]
> > > If user set command line option, then it is used:
> > > See https://wandbox.org/permlink/NIC74audyQE2ImMl
> > > Be careful, there is Runtime options: text box on the left of the code.
> > 
> > 
> > But this means you always have to specify the port, to enable TCP protocol
> 
> Yes. It is no problem.
> User should provide TCP port every time. It is better than unexpected port opens silently.
> If user don't want to provide command line options, user can use `broker.conf`.
> And `broker.conf` is provided by default.
> So user can simple run `./broker` then port 1883 is opened as TCP port.

May be we need config file not found case process:
https://wandbox.org/permlink/THnT9s38bivxN63P

**Comment by kleunen** [comment]
> TCP	WS
> NON TLS	tcp.port	ws.port
> TLS	tls.port	wss.port
> There is no `mqtt` information. It is redundant because the broker is for mqtt.
> tcp, tls, ws, and wss are similar level.

Yes I think this is best option

File: CMakeLists.txt (hunk 1 above)
Comment (redboltz) [reply CMakeLists.txt:135]: Do you move this to other PR ? Or is it required to pass CI ?

File: test/system/test_server_tls.hpp [hunk 2]
```diff
@@ -7,29 +7,30 @@
...
 public:
-    test_server_tls(as::io_context& ioc, MQTT_NS::broker::broker_t& b)
+    test_server_tls(as::io_context& ioc, boost::asio::ssl::context&& ctx, MQTT_NS::broker::broker_t& b, uint16_t port = broker_tls_port)
         : server_(
             as::ip::tcp::endpoint(
-                as::ip::tcp::v4(), broker_tls_port
+                as::ip::tcp::v4(), port
             ),
-            std::move(ctx),
+            std::forward<boost::asio::ssl::context>(ctx),
```
Comment (redboltz) [inline test/system/test_server_tls.hpp:33]:
```suggestion
            MQTT_NS::force_move(ctx),
```

ctx is rvalue-reference not forwarding reference in your code.

File: test/system/test_server_tls_ws.hpp [hunk 3]
```diff
@@ -17,20 +19,19 @@
...
-    test_server_tls_ws(as::io_context& ioc, MQTT_NS::broker::broker_t& b)
+    test_server_tls_ws(as::io_context& ioc, boost::asio::ssl::context&& ctx, MQTT_NS::broker::broker_t& b, uint16_t port = broker_tls_ws_port)
         : server_(
             as::ip::tcp::endpoint(
                 as::ip::tcp::v4(),
-                broker_tls_ws_port
+                port
             ),
-            std::move(ctx),
+            std::forward<boost::asio::ssl::context>(ctx),
```
Comment (redboltz) [inline test/system/test_server_tls_ws.hpp:34]:
```suggestion
            MQTT_NS::force_move(ctx),
```

ctx is rvalue-reference not forwarding reference in your code.

**Comment by kleunen** [comment]
I think I changed spaces to tabs now, the coding rules does not says tabs vs spaces ?

**Comment by redboltz** [comment]
> I think I changed spaces to tabs now, the coding rules does not says tabs vs spaces ?

I will add this. Please use spaces.

**Comment by kleunen** [comment]
Error copying file "/Users/runner/work/mqtt_cpp/mqtt_cpp/example/server.key.pem" to "/Users/runner/work/_temp/example/broker.key.pem".

on macos. 

Strange ...

**Comment by kleunen** [comment]
Maybe include a systemd init script for running the broker as well ?

File: CMakeLists.txt (hunk 1 above)
Comment (kleunen) [reply CMakeLists.txt]: Is moved to other PR

**Comment by redboltz** [comment]
Sorry for my late merge. I'm under winter holidays :)

**Pull request merged**